### ISSUE TITLE
Shorthand for piping a value through a bunch of functions

### ISSUE THREAD

**Issue by tfga** [jashkenas/underscore#2809]
Let's say you want to pipe some value `a` through 3 functions `f1`, `f2`, `f3`, in that order. How would you do it?

Well, you can do it the "normal" way (pure JS):

```javascript
f3(f2(f1(a)))
```

Or you can use `_.compose()`:

```javascript
_.compose( f3
         , f2
         , f1
         )(a)
```

But what I'd _really like_ to write is:

```javascript
_.pipe( a
      , f1
      , f2
      , f3
      )
```

Both JS's function composition syntax and `_.compose()` make me write the function list in the reverse order. Clojure has [threading macros](https://clojure.org/guides/threading_macros); in Elm, you could write this using the reverse apply operator:

```elm
a
|> f1
|> f2
|> f3
```

This `_.pipe()` function I'm suggesting would be a way of doing something similar in JS.

**Comment by tfga** [comment]
Somebody implemented this as a babel macro:

https://github.com/Andarist/pipeline.macro

**Comment by jgonggrijp** [comment]
@tfga You can easily implement this yourself and then add it to Underscore if you want:

```js
var pipe = _.restArguments(function(input, functions) {
    return _.compose.apply(null, functions.reverse())(input);
});

_.mixin({pipe: pipe});

_.pipe(a, f1, f2, f3);
```

Or with ES6 syntax:

```js
function pipe(input, ...functions) {
    functions.reverse();
    return _.compose(...functions)(input);
}

_.mixin({pipe});

_.pipe(a, f1, f2, f3);
```

**Comment by tfga** [comment]
@jgonggrijp But then I'd have to duplicate this code in every project. To prevent that, isn't it the point of libraries like underscore?

**Comment by jgonggrijp** [comment]
No, you don't have to copy it in every project. You can make your own NPM package that imports Underscore, adds your desired functions to it with `_.mixin`, and then exports the result as a "super Underscore". If you use that package everywhere instead of Underscore itself, no code gets duplicated and you get all original functions of Underscore unmodified as well as your own additions. Win-win!

**Comment by tfga** [comment]
@jgonggrijp Ok. Thanks for the replies.

**Comment by jgonggrijp** [comment]
@tfga I guess I should have started by mentioning why I think this doesn't belong in Underscore. Sorry for not doing that.

I see the added value of `_.pipe`. It also combines well with chaining:

```js
_.chain(a).pipe(f1, f2, f3).pipe(f4, f5).value();
```

But the problem is, there is an infinite space of functions that are not in Underscore but that would have an added value. We can't add all of them. In fact, we have to be conservative because Underscore is trying not to grow fatter (see #2060).

I'm not ruling out that new functions would be added (in fact that's not up to me to decide), but given that `pipe` is just `reverse` + `compose`, to me it doesn't seem like a game-changing addition.

Another thing I failed to mention is that there is a very similar function available from Underscore-Contrib: [`pipeline`](http://documentcloud.github.io/underscore-contrib/#pipeline).

**Comment by tfga** [comment]
@jgonggrijp No worries. Thank you for taking the time to explain this. I appreciate it.

**Comment by jgonggrijp** [comment]
Reopening this, because I'm seriously considering to replace `chain` by `pipe` in future Underscore 2.0. Motivation:

- `chain` only works with functions that have been added through `mixin`, while `pipe` can work with any function if implemented well.
- `chain` prevents treeshaking (because of `mixin`), `pipe` doesn't.

Ditching `chain` might also imply ditching the OOP notation and fully committing to functional notation instead. Discussion welcome.

FYI @tfga.

**Comment by jgonggrijp** [comment]
I would be interested in feedback from anyone who is watching, as well as @m90 and @GammaGames, on the idea of replacing `chain` by `pipe` in a future Underscore 2.0. It would still be possible to use `chain` for those who want it, but the methods wouldn't be mixed into `_` by default. In other words, people would need a line like `_ = createWrapper(allOfUnderscore)` at the top of their code in order to continue using the OOP and chaining notations.

**Comment by GammaGames** [comment]
I never use `chain` because I always found its syntax annoying to edit and retrieving the end value unnecessary. I've always treated `_` as more functional anyway since I rely on backbone for the OOP stuff, and that implementation of `pipe` would clean up some of my code so I'm all for it. Especially if I could do something like `MyBackboneCollection.pipe(_.filter, _.uniq, _.pluck)`

**Comment by m90** [comment]
I'm a heavy user of chain since forever (took me a bit to get used to it, but now that I get it, I find it to be very nice and ergonomic), but I can totally see how pipe would be the way to go 15 years later. I.e. go ahead and drop chain, I'll happily migrate to pipe or some other approach if this is really ever needed.

**Comment by jgonggrijp** [comment]
@m90 Yes I noticed that when reading your Offen benchmark, that's why I asked your feedback specifically. I'm not feeling radical enough to completely delete `_.chain` from the codebase, but it's good to know that even a heavy user like you might be willing to transition to a newer mechanism!
